python/kiss_icp/tools/visualizer.py

```diff
@@ -20,223 +20,301 @@
-        self.keypoints = self.o3d.geometry.PointCloud()
-        self.target = self.o3d.geometry.PointCloud()
-        self.frames = []
+def next_frame_callback():
+    if Kissualizer.polyscope.imgui.Button("NEXT FRAME") or Kissualizer.polyscope.imgui.IsKeyPressed(
+        Kissualizer.polyscope.imgui.ImGuiKey_N
+    ):
+        Kissualizer.block_execution = not Kissualizer.block_execution
 
-        # Initialize visualizer
-        self.vis = self.o3d.visualization.VisualizerWithKeyCallback()
-        self._register_key_callbacks()
-        self._initialize_visualizer()
 
-        # Visualization options
-        self.render_map = True
-        self.render_source = True
-        self.render_keypoints = False
-        self.global_view = False
-        self.render_trajectory = True
-        # Cache the state of the visualizer
-        self.state = (
-            self.render_map,
-            self.render_keypoints,
-            self.render_source,
-        )
+def screenshot_callback():
+    # TODO: this is just for demo, set a more valid path
+    if Kissualizer.polyscope.imgui.Button("SCREENSHOT"):
+        image_filename = "screenshot.jpg"
+        Kissualizer.polyscope.screenshot(image_filename)
+        Kissualizer.polyscope.info(f"Screenshot save at: {image_filename}")
 
-    def update(self, source, keypoints, target_map, pose):
-        target = target_map.point_cloud()
-        self._update_geometries(source, keypoints, target, pose)
-        while self.block_vis:
-            self.vis.poll_events()
-            self.vis.update_renderer()
-            if self.play_crun:
-                break
-        self.block_vis = not self.block_vis
 
-    # Private Interaface ---------------------------------------------------------------------------
-    def _initialize_visualizer(self):
-        w_name = self.__class__.__name__
-        self.vis.create_window(window_name=w_name, width=1920, height=1080)
-        self.vis.add_geometry(self.source)
-        self.vis.add_geometry(self.keypoints)
-        self.vis.add_geometry(self.target)
-        self._set_black_background(self.vis)
-        self.vis.get_render_option().point_size = 1
-        print(
-            f"{w_name} initialized. Press:\n"
-            "\t[SPACE] to pause/start\n"
-            "\t  [ESC] to exit\n"
-            "\t    [N] to step\n"
-            "\t    [F] to toggle on/off the input cloud to the pipeline\n"
-            "\t    [K] to toggle on/off the subsbampled frame\n"
-            "\t    [M] to toggle on/off the local map\n"
-            "\t    [V] to toggle ego/global viewpoint\n"
-            "\t    [T] to toggle the trajectory view(only available in global view)\n"
-            "\t    [C] to center the viewpoint\n"
-            "\t    [W] to toggle a white background\n"
-            "\t    [B] to toggle a black background\n"
-        )
+def fps_callback():
+    if Kissualizer.play_mode:
+        total_time_s = np.sum(Kissualizer.times) * 1e-9
```

Comment:
This duplicates the fps computation from the pipeline, and the timestamps are not used anywhere else in the visualizer. How about making the `get_fps` method from the pipeline a member and only passing the fps to the visualizer?

Reply:
Or we generalize this and make it possible to pass a dict of additional data to the visualizer. This callback will then print sequentially all keys and values.

Reply:
I like the "dict" idea, I will do it!